core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/ProgressImpl.java

```diff
@@ -0,0 +1,151 @@
+package uk.ac.bbsrc.tgac.miso.core.data.workflow.impl;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.EnumType;
+import javax.persistence.Enumerated;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
+import javax.persistence.Table;
+import javax.persistence.Temporal;
+import javax.persistence.TemporalType;
+
+import com.eaglegenomics.simlims.core.User;
+
+import uk.ac.bbsrc.tgac.miso.core.data.impl.UserImpl;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.AbstractProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.ProgressStep;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow.WorkflowType;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Progress;
+
+@Entity
+@Table(name = "WorkflowProgress")
+public class ProgressImpl implements Progress {
+  @Id
+  @GeneratedValue(strategy = GenerationType.AUTO)
+  private long workflowProgressId;
+
+  @Enumerated(EnumType.STRING)
+  @Column(name = "workflowName")
+  private WorkflowType workflowType;
+
+  @ManyToOne(targetEntity = UserImpl.class)
+  @JoinColumn(name = "userId", nullable = false, updatable = false)
+  private User user;
+
+  @Temporal(TemporalType.TIMESTAMP)
+  @Column(name = "created", nullable = false, updatable = false)
+  private Date creationTime;
+
+  @Temporal(TemporalType.TIMESTAMP)
+  @Column(name = "lastModified", nullable = false)
+  private Date lastModified;
+
+  @OneToMany(targetEntity = AbstractProgressStep.class, mappedBy = "id.progress")
+  private List<ProgressStep> steps;
+
+  @Override
+  public long getId() {
+    return workflowProgressId;
+  }
+
+  @Override
+  public void setId(long id) {
+    this.workflowProgressId = id;
+  }
+
+  @Override
+  public WorkflowType getWorkflowType() {
+    return workflowType;
+  }
+
+  @Override
+  public void setWorkflowType(WorkflowType workflowType) {
+    this.workflowType = workflowType;
+  }
+
+  @Override
+  public User getUser() {
+    return user;
+  }
+
+  @Override
+  public void setUser(User user) {
+    this.user = user;
+  }
+
+  @Override
+  public Date getCreationTime() {
+    return creationTime;
+  }
+
+  @Override
+  public void setCreationTime(Date creationTime) {
+    this.creationTime = creationTime;
+  }
+
+  @Override
+  public Date getLastModified() {
+    return lastModified;
+  }
+
+  @Override
+  public void setLastModified(Date lastModified) {
+    this.lastModified = lastModified;
+  }
+
+  @Override
+  public List<ProgressStep> getSteps() {
+    return steps;
+  }
+
+  @Override
+  public void setSteps(List<ProgressStep> steps) {
+    Collections.sort(steps);
```

Comment:
This works until someone uses `getSteps().add(...)`. `steps` needs to either be a sorted collection such as `TreeSet`, return immutable copy from getter, or sort on `get`. Sorted collection is probably ideal

Reply:
`steps` is now stored with a `TreeSet` implementation.

Inadvertently, this requires returning a `new` `ArrayList` from `getSteps`, so using:
```
getSteps().add(...)
```
will not modify `steps` anymore.

I've added some tests into `core` to ensure this behaviour.